snuba/clusters/cluster.py

```diff
@@ -96,14 +107,42 @@ def get_writer(
 class ClickhouseCluster(Cluster[SqlQuery, ClickhouseWriterOptions]):
     """
     ClickhouseCluster provides a reader, writer and Clickhouse connections that are
-    shared by all storages located on the cluster
+    shared by all storages located on the cluster.
+
+    ClickhouseCluster is initialized with a single address (host/port/http_port),
+    which is used for all read and write operations related to the cluster. This
+    address can refer to either the address of the actual ClickHouse server, or a
+    proxy server (e.g. for load balancing).
+
+    However there are other operations (like some DDL operations) that must be executed
+    on each individual server node, as well as each distributed table node if there
+    are multiple. If we are operating a single node cluster, this is straightforward
+    since there is only one server on which to run our command and no distributed table.
+    If we are operating a multi node cluster we need to know the full set of shards
+    and replicas on which to run our commands. This is provided by the `get_local_nodes()`
+    and `get_distributed_nodes()` methods.
     """
 
-    def __init__(self, host: str, port: int, http_port: int, storage_sets: Set[str]):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        http_port: int,
+        storage_sets: Set[str],
+        single_node: bool,
+        # The cluster name and distributed cluster name only apply if single_node is set to False
+        cluster_name: Optional[str] = None,
+        distributed_cluster_name: Optional[str] = None,
+    ):
         super().__init__(storage_sets)
+        if not single_node:
+            assert cluster_name
         self.__host = host
         self.__port = port
         self.__http_port = http_port
+        self.__single_node = single_node
+        self.__cluster_name = cluster_name
+        self.__distributed_cluster_name = distributed_cluster_name
```

Comment:
I am not sure I get why we need two cluster names since we have the single_node field. I was proposing alternatives to remove the single_node flag. But if we have it, can't we just have one cluster name ?

Reply:
Sorry if this part wasn't clear earlier. The `cluster_name` only gives the list of local servers. We'll run the local ddl on those machines. The second `distributed_cluster name` gives the list of hosts with the distributed tables. (this second part is new and needs to be added in Sentry prod) We'll run the distributed ddl on those machines. Does this approach make sense to you?